Allow children to be a ReactNode instead of ReactElement

jeffvandyke opened the pull request:
Since children is just returned by this component, there's no practical reason
that children has to be a single element. We've used multiple children in our
project, and this type definition made an upgrade incompatible.

BrianMitchL commented:
With v4, we switched to rendering a function component in order to use hooks. Unlike a class Component, a function component needs to return a JSX Element. To switch to a ReactNode, we'd need to wrap the render return with an element or a React Fragment. There are more details here: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18051

While I would rather not break the children type, I don't necessarily like wrapping what a consumer passed to children. I would rather let the consumer control exactly what is rendered  by wrapping their children in whatever markup or Fragment they'd like.

After looking into it more, it seems that the difference in types is solely a TypeScript thing, but there for historical reasons, so I'm not sure what's best. There's a good explanation here: https://stackoverflow.com/a/59840095/3928053

If you come up with a solution that passes type checking and the tests pass without changing them, I'd be happy to merge.

jeffvandyke commented:
Ahh, that makes sense :) I'll do some more investigating...

jeffvandyke commented:
Long story short - based on Flow's docs and React's behavior, DefinitelyTyped's typings for React.FunctionComponent's return (`ReactElement | null`) seem incorrect (probably should be `ReactNode`). I'll prepare and submit a PR to them, hold this PR on that, I'll post updates here.

BrianMitchL commented:
There's been an issue open about that for a few years https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18051

jeffvandyke commented:
OK, that was more complicated than I thought, so I've updated this PR to use the casting children as ReactElement workaround, seems better than introducing another Fragment layer to the component tree, and it's a workaround that's been done before.

Also updated README.md like I should have before.

giogonzo commented:
released as `v4.0.2`
